src/SqlAsyncEnumerable.cs

```diff
@@ -60,6 +61,24 @@ public SqlAsyncEnumerator(SqlConnection connection, SqlAttribute attribute)
             {
                 this._connection = connection ?? throw new ArgumentNullException(nameof(connection));
                 this._attribute = attribute ?? throw new ArgumentNullException(nameof(attribute));
+                Task.Run(() => this.GetDataAsync()).Wait();
+            }
+
+            /// <summary>
+            /// Gets the data from server into the reader ready for use.
+            /// </summary>
+            private async void GetDataAsync()
+            {
+                if (this._reader == null)
+                {
+                    using (SqlCommand command = SqlBindingUtilities.BuildCommand(this._attribute, this._connection))
+                    {
+                        await command.Connection.OpenAsync();
```

Comment:
1. We don't need to create the command here. Just open the connection. The command execution can still stay where it was
2. I'd prefer the TrackConvert be done in SqlConverters, see my other comment
3. You can just use Connection.Open() to open the connection - we don't really need to use the async version here for that.

Reply:
The idea with not executing the command is that we still want to be as lazy as possible to avoid overhead when we can. Opening the connection is a separate step from executing the command so we can just delay that until we actually need to fetch the rows - that still allows us to get server metadata like the version from the connection itself before we actually execute the command.